docs/basics/dataframe_conversion.md

```diff
@@ -0,0 +1,33 @@
+# Lightweight DataFrame conversion
+
+Some library maintainers must apply complex data frame operations, using methods and functions that may not (yet) be implemented in Narwhals. In such cases, Narwhals can still be highly beneficial, by allowing easy data frame conversion.
+
+Imagine that you maintain a library with a function that operates on Pandas data frames to produce automated reports. You want to allow users to supply data frames in any format to that function (Pandas, Polars, DuckDB, CuDF, Modin, etc.) without adding all those dependencies to your own project.
+
+One solution is to use Narwhals as a thin "DataFrame ingestion" layer, to convert user-supplied data frames to the format that your library uses internally. Since Narwhals is 0-dependency, this is a much more lightweight solution than including all the data frame libraries as dependencies.
+
+For example, this function can ingest any data frame type supported by Narwhals, and convert it to a Pandas DataFrame for internal use:
+
+```python
+import narwhals as nw
+
+
+def df_to_pandas(df):
+    out = nw.from_native(df)
+    out = out.to_pandas()
+    return out
+```
+
+Similarly, if your library uses Polars internally, you can convert any user-supplied data frames to Polars format using Narwhals.
+
+```python
+import narwhals as nw
+import polars as pl
+
+
+def df_to_polars(df):
+    out = nw.from_native(df)
+    out = out.to_arrow()
+    return pl.DataFrame(out)
```

Comment:
can this just be
```python
def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
    return nw.from_arrow(nw.from_native(df), native_namespace=pl)
```
?

Then it'll use the PyCapsule Interface, and depending on the input can avoid going via PyArrow completely

Reply:
This does not appear to work for DuckDB input, which defeats the purpose of a "universal conversion vignette."

My example first requests I call `.to_native()`. When I do that, I get an error.

```python
import narwhals as nw
from narwhals.typing import IntoDataFrame
import duckdb
import polars as pl

df = pl.DataFrame(
  {
    "A": [1, 2, 3, 4, 5],
    "fruits": ["banana", "banana", "apple", "apple", "banana"],
    "B": [5, 4, 3, 2, 1],
    "cars": ["beetle", "audi", "beetle", "beetle", "beetle"],
  }
)
df_duckdb = duckdb.sql("SELECT * FROM df")

def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
  return nw.from_arrow(nw.from_native(df), native_namespace=pl)

df_to_polars(df_duckdb).to_native()
```


Error: 


```python

In [33]: df_to_polars(df_duckdb).to_native()
---------------------------------------------------------------------------
InvalidInputException                     Traceback (most recent call last)
Cell In[33], line 1
----> 1 df_to_polars(df_duckdb).to_native()

Cell In[32], line 2, in df_to_polars(df)
      1 def df_to_polars(df: IntoDataFrame) -> pl.DataFrame:
----> 2   return nw.from_arrow(nw.from_native(df), native_namespace=pl)

File ~/Downloads/narwhals/narwhals/functions.py:473, in from_arrow(native_frame, native_namespace)
    468 implementation = Implementation.from_native_namespace(native_namespace)
    470 if implementation is Implementation.POLARS and parse_version(
    471     native_namespace.__version__
    472 ) >= (1, 3):
--> 473     native_frame = native_namespace.DataFrame(native_frame)
    474 elif implementation in {
    475     Implementation.PANDAS,
    476     Implementation.MODIN,
   (...)
    480     # These don't (yet?) support the PyCapsule Interface for import
    481     # so we go via PyArrow
    482     try:

File ~/.virtualenvs/global/lib/python3.12/site-packages/polars/dataframe/frame.py:430, in DataFrame.__init__(self, data, schema, schema_overrides, strict, orient, infer_schema_length, nan_to_null)
    426 elif hasattr(data, "__arrow_c_stream__"):
    427     # This uses the fact that PySeries.from_arrow_c_stream will create a
    428     # struct-typed Series. Then we unpack that to a DataFrame.
    429     tmp_col_name = ""
--> 430     s = wrap_s(PySeries.from_arrow_c_stream(data))
    431     self._df = s.to_frame(tmp_col_name).unnest(tmp_col_name)._df
    433 else:

File ~/Downloads/narwhals/narwhals/dataframe.py:376, in DataFrame.__arrow_c_stream__(self, requested_schema)
    374 native_frame = self._compliant_frame._native_frame
    375 if hasattr(native_frame, "__arrow_c_stream__"):
--> 376     return native_frame.__arrow_c_stream__(requested_schema=requested_schema)
    377 try:
    378     import pyarrow as pa  # ignore-banned-import

InvalidInputException: Invalid Input Error: There is no query result
```